custom/sensors/EncoderPair.java

```diff
@@ -0,0 +1,289 @@
+	 *
+	 * @param encoders
+	 *        The encoders to amalgamate.
+	 */
+	public EncoderPair(CustomEncoder encoder1, CustomEncoder encoder2) {
+		this(encoder1, encoder2, EncoderPair.DEFAULT_DISTANCE_TOLERANCE, EncoderPair.DEFAULT_RATE_TOLERANCE);
+	}
+	
+	@Override
+	public PIDSourceType getPIDSourceType() {
+		return pidSource;
+	}
+	
+	@Override
+	public double pidGet() {
+		if (pidSource == PIDSourceType.kDisplacement) {
+			return getDistance();
+		}
+		return getRate();
+	}
+	
+	@Override
+	public double pidGetSafely() throws InvalidSensorException {
+		if (pidSource == PIDSourceType.kDisplacement) {
+			return getDistanceSafely();
+		}
+		return getRateSafely();
+	}
+	
+	@Override
+	public void setPIDSourceType(PIDSourceType pidSource) {
+		if (pidSource != null) {
+			this.pidSource = pidSource;
+		}
+	}
+	
+	@Override
+	public int getSafely() throws InvalidSensorException {
+		return (int) Math.round(((double) (encoders[0].getSafely() - offset[0]) + (double) (encoders[1].getSafely() - offset[1])) / 2.0);
+	}
+	
+	@Override
+	public int get() {
+		try {
+			return getSafely();
+		}
+		catch (Exception e) {
+			LogKitten.ex(e);
+			return 0;
+		}
+	}
+	
+	@Override
+	public double getDistanceSafely() throws InvalidSensorException {
+		return ((encoders[0].getDistanceSafely() - offset[0] * encoders[0].getDistancePerPulse()) + (encoders[1].getDistanceSafely() - offset[1] * encoders[1].getDistancePerPulse())) / 2.0;
+	}
+	
+	@Override
+	public double getDistance() {
+		try {
+			return getDistanceSafely();
+		}
+		catch (Exception e) {
+			LogKitten.ex(e);
+			return 0;
+		}
+	}
+	
+	@Override
+	public boolean getDirection() {
+		return (getRate() > 0);
+	}
+	
+	@Override
+	public boolean getDirectionSafely() throws InvalidSensorException {
+		return (getRateSafely() > 0);
+	}
+	
+	@Override
+	public boolean getStopped() {
+		return Util.isZero(getRate());
+	}
+
+	@Override
+	public boolean getStoppedSafely() throws InvalidSensorException {
+		return Util.isZero(getRateSafely());
+	}
+	
+	@Override
+	public double getRateSafely() throws InvalidSensorException {
+		return (encoders[0].getRateSafely() + encoders[1].getRateSafely()) / 2.0;
+	}
+	
+	@Override
+	public double getRate() {
+		try {
+			return getRateSafely();
+		}
+		catch (Exception e) {
+			LogKitten.ex(e);
+			return 0;
+		}
+	}
+	
+	/**
+	 * Get whether this entire encoder is inverted.
+	 *
+	 * @return isInverted
+	 *         The state of inversion, true is inverted.
+	 */
+	@Override
+	public boolean getReverseDirection() {
+		return reverseDirection;
+	}
+	
+	/**
+	 * Sets the direction inversion of all encoder substituents.
+	 * This respects the original inversion state of each encoder when constructed,
+	 * and will only invert encoders if this.getReverseDirection() != the input.
+	 *
+	 * @param reverseDirection
+	 *        The state of inversion, true is inverted.
+	 */
+	@Override
+	public void setReverseDirection(boolean reverseDirection) {
```

Comment:
Unclear naming, local variable hides a field, using this. For differentiation is confusing